### PR TITLE
[4.0] Controller classes improvement

### PR DESCRIPTION
**Pull request by joomdonation**
Pull Request for Issue # .

### Summary of Changes
This PR contains improvement + clean up to controller class to support both namespace and none namespace components in Joomla 4 (backward compatible change). Below are the list of changes/improvements:

1. **Change to base Controller class**
- Allow inject Input data instead of coupling with application input
- Support creating namespace model/view classes
- Simplify getName() method (that method was implemented wrong from beginning, instead of returning name of controller, it return component name ,without com_ prefix), just use substr method instead of regex.
- Implement getControllerName() method to allow getting name of controller properly (as mentioned above, that should be the role of getName() method)

2. **Change to Form Controller class**
- Call getControllerName method (from base controller class) to get context instead of using regex
- Use toPlural method of \Joomla\String\Inflector class to detect view_list property instead of using a script found elsewhere in the internet.
- Remove some redundant else if

3. **Change to Admin Controller class**
- Use controller input instead of application input to get input data (this was fixed in staging but not updated to 4.0 branch for some reasons)
- Implement getModel method. That will help us not having to add blank method like this https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_content/controllers/articles.php#L121 in every admin controller

### Testing Instructions

1. Install Joomla 4.0
2. Apply patch
3. Try to navigate to different pages (maybe try to add. edit, delete, publish articles), make sure it is still working as before

I also tested with the namespace com_content component in this PR https://github.com/joomla/joomla-cms/pull/14370 and it worked well. Hard for end user to test it for now, thought

Need developers to review code and discuss, too

### CODE REVIEW

File: libraries/src/Cms/Controller/Controller.php [hunk 1]
```diff
@@ -323,30 +339,38 @@ public static function getInstance($prefix, $config = array())
...
 		{
 			\JLog::addLogger(array('text_file' => 'jcontroller.log.php'), \JLog::ALL, array('controller'));
 		}
 
-		$this->input = \JFactory::getApplication()->input;
+		$this->input  = $input ? $input : \JFactory::getApplication()->input;
+		$this->option = $this->input->getCmd('option');
+
+		// Check to make sure the component is enabled
+		if (!\JComponentHelper::isEnabled($this->option))
```
Comment (mbabker) [inline libraries/src/Cms/Controller/Controller.php]: Why?

File: libraries/src/Cms/Controller/Controller.php [hunk 2]
```diff
@@ -541,9 +568,26 @@ protected function checkEditId($context, $id)
...
+		$config['option'] = $this->option;
+
+		// If this is a namespace controller, create namespace model class
+		if ($this->namespace)
+		{
+			$modelClass = $this->namespace . '\\Model\\' . ucfirst($name);
+
+			if (class_exists($modelClass))
+			{
+				return new $modelClass($config);
```
Comment (mbabker) [inline libraries/src/Cms/Controller/Controller.php]: Why direct instantiation instead of going through the existing factory method?

File: libraries/src/Cms/Controller/Controller.php (hunk 1 above)
Comment (joomdonation) [inline libraries/src/Cms/Controller/Controller.php]: If component is not enabled, why we would we allow creating a controller belong to this component? Or I mis-understand something?

Comment (mbabker) [inline libraries/src/Cms/Controller/Controller.php]: I don't think we need to worry in the MVC layer about respecting that state personally.  For me it's more important the state be respected in things like routing, but why should we in effect block someone from being able to create and use a controller for a disabled component?

Comment (joomdonation) [inline libraries/src/Cms/Controller/Controller.php]: Honestly, I could not think of a case we allow executing controller of a blocked component. If there is a reason for removing that check, I think we can agree that this should be the role of a component dispatcher (or JComponentHelper right now). So if you think it is OK to remove it, then I can remove it. Make the code even cleaner :)

Comment (mbabker) [inline libraries/src/Cms/Controller/Controller.php]: Do this as a separate PR so it can get the attention it deserves.  But to me, I don't think the controller constructors need to have this check.

Comment (joomdonation) [inline libraries/src/Cms/Controller/Controller.php]: OK. I removed it for now. We will come back to this later if needed.

File: libraries/src/Cms/Controller/Controller.php (hunk 2 above)
Comment (joomdonation) [inline libraries/src/Cms/Controller/Controller.php]: Thanks. I updated code to use Model::getInstance() . Should be OK now.

File: libraries/src/Cms/Controller/Controller.php [hunk 3]
```diff
@@ -323,30 +339,32 @@ public static function getInstance($prefix, $config = array())
 	 * @param   array  $config  An optional associative array of configuration settings.
 	 * Recognized key values include 'name', 'default_task', 'model_path', and
 	 * 'view_path' (this list is not meant to be comprehensive).
+	 * @param   Input  $input   The controller input
 	 *
 	 * @since   3.0
 	 */
-	public function __construct($config = array())
+	public function __construct($config = array(), Input $input = null)
```
Comment (laoneo) [inline libraries/src/Cms/Controller/Controller.php]: Why not injecting here the application, like that we can remove a lot of JFactory::get* calls. Also JText::_ can then be done trough the app->getLanguage()->_ call.

Comment (joomdonation) [inline libraries/src/Cms/Controller/Controller.php]: I can do that. I want to ask for feedback from others before making the change. How do you think about it @mbabker , @wilsonge , @yvesh?

Comment (mbabker) [inline libraries/src/Cms/Controller/Controller.php]: `JText::_()` does a couple extra things that `JLanguage::_()` does not do, I wouldn't encourage going this route.

File: libraries/src/Cms/Controller/Admin.php [hunk 4]
```diff
@@ -71,7 +71,14 @@ public function __construct($config = array())
 		// Guess the \JText message prefix. Defaults to the option.
 		if (empty($this->text_prefix))
 		{
-			$this->text_prefix = strtoupper($this->option);
+			if (isset($config['text_prefix']))
```
Comment (yvesh) [inline libraries/src/Cms/Controller/Admin.php]: empty please :-)

**Comment by joomdonation** [comment]
I have one question here. Is it safe to assume that in context of Joomla Controller, a controller always has input object available? And we can safely get $option directly from Input?

I ask this question is because unit test is failed for this PR. One of the test is validating detect $option from class name

```php
$object = new MincesControllerMince(
	array(
		// Neutralise a JPATH_COMPONENT not defined error.
		'base_path' => JPATH_BASE . '/component/com_foobar'
	)
);
```

Do we have to implement code to pass this test? Mean when $option is empty from input, we need to detect $option from class name (com_minces in this case)?

File: libraries/src/Cms/Controller/Admin.php (hunk 4 above)
Comment (joomdonation) [inline libraries/src/Cms/Controller/Admin.php]: It makes sense to use **1empty**. However, there are many other places in the class (and other classes, I think), we are using **array_key_exists** . I guess we need a decision about which one is best before making the change

Comment (yvesh) [inline libraries/src/Cms/Controller/Admin.php]: @mbabker @laoneo @wilsonge thoughts?

File: libraries/src/Cms/Controller/Controller.php (hunk 3 above)
Comment (laoneo) [inline libraries/src/Cms/Controller/Controller.php]: Somehow I would like to get a way around all this static calls. In the Dispatcher we are using the language from the app to translate strings.

File: libraries/src/Cms/Controller/Admin.php (hunk 4 above)
Comment (laoneo) [inline libraries/src/Cms/Controller/Admin.php]: IMO it must not be consistent at all, just correct when used. In that case makes sense to use `empty` as it can be set but empty. So the actual code would let it trough, what we don't want.

File: libraries/src/Cms/Controller/Controller.php (hunk 3 above)
Comment (mbabker) [inline libraries/src/Cms/Controller/Controller.php]: You would have to turn JText into a non-static helper, not avoid using it by digging into the JLanguage.  The two are not functionally similar.

**Comment by joomdonation** [comment]
At this point, the biggest concern is how we will calculate component base namespace (joomla\Component\Content\Site or Joomla\Component\Content\Admin) of the controller. This base namespace allows us to creating correct model / view classes on getModel() and getView() methods

In this PR, I have to implement getNamespace() method to get this base namespace from class name of controller. In a different PR, @laoneo use str_replace to get class name of Model and View class (basically, replace **Controller** with **Model** and/or **View**). Both the options are not really nice

One different solution would be getting namespace from database (which we discussed on a different PR).  Imagine that from $option = com_content, we can get namespace is **Joomla\Component**. Look at current getModel() method, we will have to implement some code like this

```php
public function getModel($name = '', $prefix = '', $config = array())
{
	$app = JFactory::getApplication();
	
	if ($app->isClient('site'))
	{
		$modelNamespace = $this->namespace . '\\Site';
	}
	else
	{
		$modelNamespace = $this->namespace . '\\Site\\Admin';
	}
}
```
Or if we want to allow creating both frontend and backend model, we can introduce a new parameter to the method $section = 'auto'

```php
public function getModel($name = '', $prefix = '', $config = array(), $section = 'auto')
{
	$section = ucfirst($section);
	
	if ($section == 'Site')
	{
		$modelNamespace = $this->namespace . '\\Site';
	}
	elseif ($section == 'Admin')
	{
		$modelNamespace = $this->namespace . '\\Site';
	}
	else
	{
		$app = JFactory::getApplication();
	
		if ($app->isClient('site'))
		{
			$modelNamespace = $this->namespace . '\\Site';
		}
		else
		{
			$modelNamespace = $this->namespace . '\\Site\\Admin';
		}	
	}
}
```
Will this option works? And hopefully, better?

**Comment by joomdonation** [comment]
So I changed the code so that the namespace is not getting from manifest cache instead of calculating from class name using reflection.  If we are worry about the size of manifest cache, we can introduce a new column to store namespace of extension if needed.

The $prefix parameter or **getModel()** and **getView()** method now can has two special value Site/Admin and the  controller will create Site or Admin Model/View class. In case it is empty, the current context (depend on application) will be used

@mbabker Could you please take a quick review to see whether it is OK to have it implemented like this?

**Comment by joomdonation** [comment]
Does anyone know why Travis throws this error for this PR and how to correct it?

> 1) JControllerFormTest::testConstructor
> JDatabaseExceptionConnecting: Could not connect to MySQL.

**Comment by mbabker** [comment]
The test classes only set up a mocked database connection if they extend `TestCaseDatabase`, or you've explicitly injected a mock database driver into `JFactory::$database`.  So the test class now probably has no database dependencies but your changes are adding one.

File: libraries/src/Cms/Controller/Controller.php [hunk 5]
```diff
@@ -404,12 +440,48 @@ public function __construct($config = array())
...
+		if (empty($this->viewPrefix))
+		{
+			if (array_key_exists('view_prefix', $config))
+			{
+				// User-defined prefix
+				$this->viewPrefix = $config['view_prefix'];
+			}
+			elseif ($this->namespace)
+			{
+				if ($app->isClient('site'))
```
Comment (yvesh) [inline libraries/src/Cms/Controller/Controller.php]: There is also client installation etc. basically it can be any because it's the name because you can set it when you extend JApplication.

One genral note: For me we coupling again the controller to tight to the view..

Comment (mbabker) [inline libraries/src/Cms/Controller/Controller.php]: Until you get away from the controller acting as its own factory for creating class objects (the fact it is aware of model and view class names is the indicator of this problem), this level of coupling is going to exist in the system.

**Comment by joomdonation** [comment]
With latest change, I get namespace of component from manifest cache (instead of getting it from class name) as suggested somewhere, maybe that's why database is needed now

Since I don't have any experience with unittest, any instructions to correct this?

**Comment by mbabker** [comment]
Start by ensuring `JControllerFormTest extends TestCaseDatabase` and we'll go from there.

**Comment by joomdonation** [comment]
Great. It worked, also for my other PR https://github.com/joomla/joomla-cms/pull/14567 . Thanks Michael

To me, right now, there are some small concern left for this PR:

1. Should we get namespace of extensions from manifest cache as how I am doing in this PR or we should introduce a new field to #__extensions table to store namespace. Manifest cache contains much more data than we need, and we are loading all components into memory, so using Manifest cache is of course expensive than using a separate field

2. The name of controller classes. At the moment, it is Controller, Form, Admin. Should we change it to different names like BaseController, FormController, AdminController (like how classes are named in some other framework packages). I don't know what's the best practive

Other than that, I think I am done for changing Controller classes to support namespace components. I also opened another PR https://github.com/joomla/joomla-cms/pull/14567 for changing Model classes, too and I think I am done for that part, too

Next thing would be the View classes (change it to support namespace, split the View classes to View and Html....to support more format in the future). However, before working on it, I would like to see feedback from the Joomla 4 team to see whether I am on right direction or not. So @mbabker @wilsonge @yvesh @laoneo please help reviewing this PR and https://github.com/joomla/joomla-cms/pull/14567 when you have some time.

**Comment by laoneo** [comment]
Honestly for something doesn't sound right when we need database based tests when I just want to test the controller.

**Comment by laoneo** [comment]
@joomdonation I think it would make sense to have the com_content pr merged first. Like that we can see how it will be used and we can merge your pr's.

**Comment by joomdonation** [comment]
> Honestly something doesn't sound right for me when we need database based tests when I just want to test the controller.

I think the Joomla 4 team will need to decide where we get / store namespace of extension first. I remember on different PRs, @wilsonge and @mbabker mentioned that can store it in database (a new field or in manifest cache) to have have a single source to get namespace. That's the reason I made the change.

**Comment by joomdonation** [comment]
> @joomdonation I think it would make sense to have the com_content pr merged first. Like that we can see how it will be used and we can merge your pr's.

One thing with merging that PR is it seems it makes cpanel of Joomla4 installation broken (maybe I am wrong), downloaded your repo for testing long time ago and made several changes while testing my PR (I uses your namespaced com_content for testing)

**Comment by mbabker** [comment]
Part of the problem we have right now is the MVC layer isn't doing dependency injection of anything it needs, so the base classes themselves are doing service/parameter lookups.  We really need to shift the creation behaviors into a proper factory object.

Ya, that means there's still a dependency to an external data provider somewhere, but we're isolating that dependency to a single `Joomla\Cms\MVC\Factory` type object instead of all over the place.

**Comment by laoneo** [comment]
We have been talking about that in London already. I agree that a factory would solve that issue. But I think that switch can be done on a later stage. For me it is more important to go a step further and merge the PR's that we can star thinking about such stuff. IMO we shouldn't do all of the issues we are discussing here in this pr. Let's go step by step.

@joomdonation on some point we need to make that step and for the current state, I don't care if Joomla 4 can be installed trough cPanel. I guess now it is important to move further and not end up in endless discussions.

**Comment by joomdonation** [comment]
@laoneo To be more clear (as I said, maybe my installation was broken because I made some changes to it to be able to test your namespace com_content with my PR), when I login to administrator area http://localhost/namespace/administrator/, I see this error

error_administrator

So if you confirm that you have no error like that, it would be OK.

As we have testers testing others part of Joomla 4, I think we cannot merge PR which result the error like this. If we want to merge the namespace com_content PR, we will have to make sure nothing is broken (maybe there is some modules in backend use com_content models but the none-namespace one and it causes the errors)

**Comment by laoneo** [comment]
Yes for sure, it must not be broken :smile: actually there are also some merge conflicts. But as soon as I have the go from the J4 release leader ( @wilsonge ) then I will solve the merge conflicts.

**Comment by joomdonation** [comment]
Maybe I should still go ahead to work on some changes to View class to support namespace component. Then my PRs even is being merged before your namespace com_content PR, who knows, lol

@mbabker Could you please give more detailed instructions about how should we implement Joomla\Cms\MVC\Factory ? What dependency should be passed to that class? This is honestly, over my skill, thus need more helps
